### ISSUE TITLE
Thread question

### ISSUE THREAD

**Issue by LKmubihei** [usnistgov/ARIAC#219]
Is there a solution to the problem I raised at the last meeting about two robots moving simultaneously?

**Comment by jaybrecht** [comment]
Please take a look at the [branch tutorial_8](https://github.com/jaybrecht/ariac_tutorials/tree/tutorial_8) on the ariac_tutorials repo. Let me know if this solves the issue. 

The instructions for running are as follows:

- Clone the repo in the src directory of your ariac_ws
- Switch to the tutorial 8 branch
- Build the workspace
- Source the install in both terminals
- Run the following commands

_in terminal 1:_
`ros2 run ariac_tutorials tutorial_8.py`

_in terminal 2:_
`ros2 launch ariac_gazebo ariac.launch.py`

The tutorial_8 executable does the following:
- starts the competition
- waits for the controllers to be active
- sends both robots to their home position
- waits for the execution of both robots to finish
- waits for a few seconds between movements 
- sends both robots to a second position
- waits for the execution of both robots to finish
- exits

**Comment by LKmubihei** [comment]
I don't think the code you provided can solve the problem. 

`    try:
        while not commander.finished_executing():
            executor.spin_once()
    except GoalRejectedException as e:
        print(e)
`

This code is not much different from the given wait() function, except that now two robots are waiting together. This is unacceptable for achieving the goal of having both robots complete their tasks at the same time. For example, if the floor robot is executing the kitting task and the ceiling robot is executing the assembly task, each robot needs to plan and execute trajectories multiple times while performing their respective tasks. If both robots execute their trajectories together, the floor robot must wait for the ceiling robot to finish executing its trajectory before starting the next step. Therefore, a better solution is to determine the wait for each robot to finish executing a trajectory separately, as implemented in this code

`

def floor_finished_executing(self):

    if self.floor_robot_goal_status == Status.REJECTED:
        raise GoalRejectedException("floor robot rejected the goal")
    if (self.floor_robot_goal_status == Status.FINISHED):
        return True
    return False
`
However, waiting separately will definitely block the thread. Since in one executor, only one spin_once() function can be executed at the same time, using executor.spin_once() makes it impossible to open a new thread to wait at the same time. I have tried using two executors to execute the spin_once() function separately, but encountered blocking when subscribing to the topic.

**Comment by jaybrecht** [comment]
My suggestion in this case would to have separate nodes for each robot and a single supervisor node all started separately through a launch file. The robot commander nodes would each run exclusively through callbacks and the supervisor node can send requests to the commander nodes through ROS topics, services, or actions. 

For example the floor robot commander node could host a service server called PickObjectFromBin. The supervisor could send a request asynchronously and the floor robot would execute all of the necessary movements then return a response when the request is completed. At the same time the supervisor could also send a request to the ceiling robot to perform another task through a service call or action goal. You would need to define all the necessary ROS messages yourself in this case but it should be doable. Trying to do everything in a single executable in different threads may be possible but will probably be much harder to organize and much easier to create a deadlock. 

As an example you can look at [tutorial_7](https://ariac.readthedocs.io/en/latest/tutorials/tutorial_7.html). There is the interface node that sends a Trigger request to the commander node to move each of the robots to a home position. In this case there is a single robot commander node but for more flexibility each robot could have its own separate commander node.

**Comment by jaybrecht** [comment]
Also once you are in the service callback inside of one of the robot commander nodes you will need to handle sending the goal to the action server to move the robot. This [link](https://answers.ros.org/question/322831/ros2-wait-for-action-using-rclpy/) shows an example for sending a ROS action inside of service callback using the Event class as the flag to wait until the action is completed. 

Note that you need to add the service server and action client to a Reentrant callback group and start the node as part of a multi-threaded executor in order for this to work properly.
